### PR TITLE
refactor covariant to use less goto

### PR DESCRIPTION
**Pull request by thewilsonator**

### CODE REVIEW

File: src/dmd/mtype.d [hunk 1]
```diff
@@ -601,77 +612,81 @@ extern (C++) abstract class Type : ASTNode
...
             if (t1.parameterList.length != t2.parameterList.length)
-                goto Ldistinct;
+                return distinct();
 
             foreach (i, fparam1; t1.parameterList)
             {
                 Parameter fparam2 = t2.parameterList[i];
-
-                if (!fparam1.type.equals(fparam2.type))
+                void cov() { notcovariant |= !fparam1.isCovariant(t1.isref, fparam2); }
```
Comment (ibuclaw) [inline src/dmd/mtype.d:628]:
```suggestion
                void checkParametersCovariant() { notcovariant |= !fparam1.isCovariant(t1.isref, fparam2); }
```

No need to keep the function name cryptic.

Comment (thewilsonator) [reply src/dmd/mtype.d]: done

File: src/dmd/mtype.d [hunk 2]
```diff
@@ -588,6 +588,17 @@ extern (C++) abstract class Type : ASTNode
             //    printf("ty = %d\n", next.ty);
             printf("mod = %x, %x\n", mod, t.mod);
         }
+        int distinct()
+        {
+            //printf("\tcovaraint: 0\n");
```
Comment (ibuclaw) [inline src/dmd/mtype.d:593]:
```suggestion
            //printf("\tcovariant: 0\n");
```

File: src/dmd/mtype.d [hunk 3]
```diff
@@ -588,6 +588,17 @@ extern (C++) abstract class Type : ASTNode
...
+            return 0;
+        }
+
+        int notCovariant()
+        {
+            //printf("\tcovaraint: 2\n");
```
Comment (ibuclaw) [inline src/dmd/mtype.d:599]:
```suggestion
            //printf("\tcovariant: 2\n");
```

File: src/dmd/mtype.d [hunk 4]
```diff
@@ -601,77 +612,81 @@ extern (C++) abstract class Type : ASTNode
...
+                {
+                    if (ts1.sym == (cast(TypeStruct)tp2).sym && MODimplicitConv(tp2.mod, tp1.mod))
+                        checkParametersCovariant();
+                    else return distinct();
+                }
+                else if (tp1.ty == Tpointer)
+                {
+                    if (tp2.implicitConvTo(tp1))
+                        checkParametersCovariant();
+                    else return distinct();
```
Comment (ibuclaw) [inline src/dmd/mtype.d:655]: Actually, couldn't these conditions be inverted, then you can have the checkParametersCovariant at the end of the scope unconditionally?


```suggestion
                    if (!tp2.implicitConvTo(tp1))
                        return distinct();
```
(then the other call goes...)

File: src/dmd/mtype.d [hunk 5]
```diff
@@ -601,77 +612,81 @@ extern (C++) abstract class Type : ASTNode
...
+                }
+                else if (tp1.ty == Tdelegate)
+                {
+                    if (tp1.implicitConvTo(tp2))
+                       checkParametersCovariant();
+                    else return distinct();
                 }
-            Lcov:
-                notcovariant |= !fparam1.isCovariant(t1.isref, fparam2);
```
Comment (ibuclaw) [inline src/dmd/mtype.d:653]: (... here)

Comment (ibuclaw) [reply src/dmd/mtype.d:653]: It looks like you are also missing the following at the end of the if/else branches:
```
else
    return distinct();
```

**Comment by 12345swordy** [comment]
Test failing, need rebase.

**Comment by RazvanN7** [comment]
ping @thewilsonator

**Comment by RazvanN7** [comment]
Tests are consistently failing.

**Comment by thewilsonator** [comment]
This is well out of date. I detect no more `goto` in `mtype.d`!

**Pull request closed without merging**